gremlin-dotnet/src/Gremlin.Net/Driver/SingleMessageResultReceiver.cs

```diff
@@ -0,0 +1,88 @@
+#region License
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Gremlin.Net.Driver.Messages;
+using Gremlin.Net.Driver.ResultsAggregation;
+using Gremlin.Net.Structure.IO.GraphSON;
+using Newtonsoft.Json.Linq;
+
+namespace Gremlin.Net.Driver
+{
+    internal class ResponseHandlerForSingleRequestMessage<T> : IResponseHandlerForSingleRequestMessage
+    {
+        public Task<ResultSet<T>> Result => _tcs.Task;
+
+        private readonly TaskCompletionSource<ResultSet<T>> _tcs =
+            new TaskCompletionSource<ResultSet<T>>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+        private readonly GraphSONReader _graphSONReader;
+        private bool _isAggregatingSideEffects;
+        private IAggregator _aggregator;
+        private readonly List<T> _result = new List<T>();
+
+        public ResponseHandlerForSingleRequestMessage(GraphSONReader graphSonReader)
+        {
+            _graphSONReader = graphSonReader;
+        }
+
+        public void HandleReceived(ResponseMessage<JToken> received)
+        {
+            var receivedData = typeof(T) == typeof(JToken)
+                ? new[] {received.Result.Data}
+                : _graphSONReader.ToObject(received.Result.Data);
```

Comment:
It would be nice to remove the dynamic call here, i.e., cast to `object`.

That way the failure would be far more simple to understand if its static code.

Reply:
You mean that `_graphSONReader.ToObject()` shouldn't return `dynamic`?

Reply:
I mean converting it to whatever you expect (`IEnumerable`?) to avoid making `_aggregator.Add(d)` and `_result.Add(d)` being dynamic calls.

Reply:
OK, then I really misunderstood your suggestion. I tried your suggestion, but unfortunately it wasn't that easy as it would require a change to this line, too:
```cs
_result.Add(d);
```
which doesn't work if `d` is of type `object`. We would need a cast to `T` here, but that would require more handling in case that `d` is not of type `T`.
So, I prefer not to change this now and instead improve type safety of the deserialization in general some time later.